core/matrix/ell.hpp

```diff
@@ -0,0 +1,268 @@
+
+    /**
+     * Returns the padding of the matrix.
+     */
+    size_type get_padding() const noexcept { return padding_; }
+
+    /**
+     * Returns a value in the ELLPACK format.
+     *
+     * @param row  the row of the requested element
+     * @param col  the column of the requested element
+     *
+     * @note  the method has to be called on the same Executor the matrix is
+     *        stored at (e.g. trying to call this method on a GPU matrix from
+     *        the CPU results in a runtime error)
+     */
+    value_type &val_at(size_type row, size_type col) noexcept
+    {
+        return values_.get_data()[linearize_index(row, col)];
+    }
+
+    /**
+     * @copydoc Ell::val_at(size_type, size_type)
+     */
+    value_type val_at(size_type row, size_type col) const noexcept
+    {
+        return values_.get_const_data()[linearize_index(row, col)];
+    }
+
+    /**
+     * Returns a column index of the ELLPACK foramt.
+     *
+     * @param row  the row of the requested element
+     * @param col  the column of the requested element
+     *
+     * @note  the method has to be called on the same Executor the matrix is
+     *        stored at (e.g. trying to call this method on a GPU matrix from
+     *        the CPU results in a runtime error)
+     */
+    index_type &col_at(size_type row, size_type col) noexcept
+    {
+        return get_col_idxs()[linearize_index(row, col)];
+    }
+
+    /**
+     * @copydoc Ell::col_at(size_type, size_type)
+     */
+    index_type col_at(size_type row, size_type col) const noexcept
+    {
+        return get_const_col_idxs()[linearize_index(row, col)];
+    }
+
+protected:
+    /**
+     * Creates an empty ELL matrix.
+     *
+     * @param exec  Executor associated to the matrix
+     */
+    explicit Ell(std::shared_ptr<const Executor> exec)
+        : BasicLinOp<Ell>(exec, 0, 0, 0),
+          values_(exec),
+          col_idxs_(exec),
+          max_nnz_row_(0),
+          padding_(0)
+    {}
+
+    /**
+     * Creates an uninitialized Ell matrix of the specified size.
+     *
+     * @param exec  Executor associated to the matrix
+     * @param num_rows      number of rows
+     * @param num_cols      number of columns
+     * @param num_nonzeros  number of nonzeros
+     * @param max_nnz_row   maximum number of nonzeros in one row
+     * @param padding       padding of the rows
+     */
+    Ell(std::shared_ptr<const Executor> exec, size_type num_rows,
+        size_type num_cols, size_type num_nonzeros, size_type max_nnz_row,
+        size_type padding)
+        : BasicLinOp<Ell>(exec, num_rows, num_cols, num_nonzeros),
+          values_(exec, num_rows*max_nnz_row),
+          col_idxs_(exec, num_rows*max_nnz_row),
+          max_nnz_row_(max_nnz_row),
+          padding_(padding)
+    {}
+
+    /**
+     * Creates an uninitialized Ell matrix of the specified size.
+     *    (When padding is not specified.)
```

Comment:
Use: "The padding is set to the number of rows of the matrix." This gives more information.

Reply:
Fixed